Tasks/Common/RemoteDeployer/JobHelper.ps1

```diff
@@ -0,0 +1,189 @@
+                    Receive-Job -Job $job |
+                        ForEach-Object {
+                            if($_.VstsTaskJobResult -eq $true) { 
+                                $jobResults += $_ 
+                            } else {
+                                if($_ -is [System.Management.Automation.ErrorRecord]) {
+                                    $errorRecord = $_
+                                    $null = & { try { & $errorHandler $errorRecord $($job.Location) } catch { Write-Host "ErrorHandlerException: $($_.Exception.ToString())" } }
+                                } else {
+                                    $outputObject = $_
+                                    $null = & { try { & $outputHandler $outputObject $($job.Location) } catch { Write-Host "OutputHandlerException: $($_.Exception.ToString())" } }
+                                }
+                            }
+                        }
+                        
+                    if($jobState -eq "completed") {
+                        $remoteExecutionStatusByLocation[$computerName] = "Finished"
+                    } elseif($jobState -ne "running") {
+                        Write-Verbose "Job (Id = $jobId) is in undesirable state (State = $jobState). Attempting reconnection"
+                        if($connectionAttemptsByLocation[$computerName] -ge 15) {
+                            Write-Verbose "Maximum connection retry limit reached for computerName: $computerName"
+                            $remoteExecutionStatusByLocation[$computerName] = "Finished"
+                        } else {
+                            $newJobId = Retry-Connection -targetMachines $targetMachines -computerName $computerName -sessionName $sessionName
+                            if($newJobId -ne $null) {
+                                Write-Verbose "Connection re-established to computer: $computerName, JobId (New): $newJobId, JobId(Old): $($jobInfo.Id)"
+                                Stop-Job -Id $jobInfo.Id
+                                $jobInfo.Id = $newJobId
+                                $connectionAttemptsByLocation[$computerName] = 0;
+                            } else {
+                                Write-Verbose "Unable to re-establish connection to computer: $computerName. Retry attempt #$($connectionAttemptsByLocation[$computerName])"
+                                $connectionAttemptsByLocation[$computerName]++;
+                            }
+                        }
+                    }
+                }
+            }
+
+            $isRemoteExecutionFinished = $true
+            foreach($value in $remoteExecutionStatusByLocation.Values) {
+                if($value -eq "Unknown") {
+                    $isRemoteExecutionFinished = $false
+                    break;
+                }
+            }
+
+            Start-Sleep -Seconds 30
```

Comment:
1. Should we reduce this to say 10?
2. Should this be inside the above block
 foreach($value in $remoteExecutionStatusByLocation.Values) { 
     if($value -eq "Unknown") {
          $isRemoteExecutionFinished = $false
              break; 
              Start-Sleep -Seconds 10
     }
}